gh-85162: Add HTTPSServer to http.server (#129607)
python/cpython: Doc/library/http.server.rst

   .. versionadded:: 3.7

> .. class:: HTTPSServer(server_address, RequestHandlerClass, \
>                        bind_and_activate=True, *, certfile, keyfile=None, \
>                        password=None, alpn_protocols=None)

Reviewer: ```suggestion
.. class:: HTTPSServer(server_address, RequestHandlerClass,\
                       bind_and_activate=True, *, certfile, keyfile=None,\
                       password=None, alpn_protocols=None)
```

>    This class is a :class:`HTTPServer` subclass with a wrapped socket using the
>    :mod:`ssl`, if the :mod:`ssl` module is not available the class will not
>    initialize.

Reviewer: ```suggestion
   Subclass of :class:`HTTPServer` with a wrapped socket using the :mod:`ssl` module.
   If the :mod:`ssl` module is not available, instantiating an :class:`!HTTPSServer`
   object fails with an :exc:`ImportError`.
```

>    The *certfile* argument is required and is the path to the SSL
>    certificate chain file. The *keyfile* is the path to its private key. But
>    private keys are often protected and wrapped with PKCS #8, so we provide
>    *password* argument for that case.

Reviewer: ```suggestion
   The *certfile* argument is the path to the SSL certificate chain file,
   and the *keyfile* is the path to file containing the private key.
   
   A *password* can be specified for files protected and wrapped with PKCS#8,
   but beware that this could possibly expose hardcoded passwords in clear.
```

>    The *alpn_protocols* argument, if provided, should be a sequence of strings
>    specifying the Application-Layer Protocol Negotiation (ALPN) protocols
>    supported by the server. ALPN allows the server and client to negotiate
>    the application protocol during the TLS handshake. By default, it is set
>    to ``["http/1.1"]``, meaning the server will support HTTP/1.1.

Reviewer: ```suggestion
   When specified, the *alpn_protocols* argument must be a sequence of strings
   specifying the "Application-Layer Protocol Negotiation" (ALPN) protocols
   supported by the server. ALPN allows the server and the client to negotiate
   the application protocol during the TLS handshake.

   By default, it is set to ``["http/1.1"]``, meaning the server supports HTTP/1.1.
```

   .. versionadded:: next

> .. class:: ThreadingHTTPSServer(server_address, RequestHandlerClass, \
>                        bind_and_activate=True, *, certfile, keyfile=None, \
>                        password=None, alpn_protocols=None)

Reviewer: ```suggestion
.. class:: ThreadingHTTPSServer(server_address, RequestHandlerClass,\
                                bind_and_activate=True, *, certfile, keyfile=None,\
                                password=None, alpn_protocols=None)
```

>    This class is identical to :class:`HTTPSServer` but uses threads to handle
>    requests by using the :class:`~socketserver.ThreadingMixIn`. This is
>    analogue of :class:`ThreadingHTTPServer` class only using
>    :class:`HTTPSServer`.

Reviewer: ```suggestion
   This class is identical to :class:`HTTPSServer` but uses threads to handle
   requests by inheriting from :class:`~socketserver.ThreadingMixIn`. This is
   analogous to :class:`ThreadingHTTPServer` only using :class:`HTTPSServer`.
```

   .. versionadded:: 3.7

.. class:: HTTPSServer(server_address, RequestHandlerClass,\
                       bind_and_activate=True, *, certfile, keyfile=None,\
                       password=None, alpn_protocols=None)

   Subclass of :class:`HTTPServer` with a wrapped socket using the :mod:`ssl` module.

>    If the :mod:`ssl` module is not available, instantiating an :class:`!HTTPSServer`

Reviewer: ```suggestion
   If the :mod:`ssl` module is not available, instantiating a :class:`!HTTPSServer`
```

I think I was wrong before. It should *a* HTTP because it's a Hypertext and not an Hypertext. My bad.

>    object fails with an :exc:`RuntimeError`.

Reviewer: ```suggestion
   object fails with a :exc:`RuntimeError`.
```

Reviewer: This one still needs the change.

   The *certfile* argument is the path to the SSL certificate chain file,
   and the *keyfile* is the path to file containing the private key.

   A *password* can be specified for files protected and wrapped with PKCS#8,

>    but beware that this could possibly expose hardcoded passwords in clear.

Reviewer: Since it's for a testing server, users may not be experimented so maybe add a link to indicate that the parmeters certfile, keyfile and password are passed to https://docs.python.org/3/library/ssl.html#ssl.SSLContext.load_cert_chain.

Author: sorry but I think this is overcomplicating, specifying which functions the parameters are passed to.

Author: this is too detailed information that is usually found in the source code

but if you send me suggestion I will accept it :)

Reviewer: It's documented in `load_cert_chain` actually what `password` is expected to be

Reviewer: It says:

> The password argument may be a function to call to get the password for decrypting the private key. It will only be called if the private key is encrypted and a password is necessary. It will be called with no arguments, and it should return a string, bytes, or bytearray. If the return value is a string it will be encoded as UTF-8 before using it to decrypt the key. Alternatively a string, bytes, or bytearray value may be supplied directly as the password argument. It will be ignored if the private key is not encrypted and no password is needed.

Reviewer: You can just use `.. seealso::`

Author: i added it

   .. versionadded:: 3.7

.. class:: HTTPSServer(server_address, RequestHandlerClass, \
                       bind_and_activate=True, *, certfile, keyfile=None, \
                       password=None, alpn_protocols=None)

   This class is a :class:`HTTPServer` subclass with a wrapped socket using the
   :mod:`ssl`, if the :mod:`ssl` module is not available the class will not
   initialize.

   The *certfile* argument is required and is the path to the SSL
   certificate chain file. The *keyfile* is the path to its private key. But
   private keys are often protected and wrapped with PKCS #8, so we provide
   *password* argument for that case.

   The *alpn_protocols* argument, if provided, should be a sequence of strings
   specifying the Application-Layer Protocol Negotiation (ALPN) protocols
   supported by the server. ALPN allows the server and client to negotiate
   the application protocol during the TLS handshake. By default, it is set
   to ``["http/1.1"]``, meaning the server will support HTTP/1.1. Other

>    possible values may include ``["h2", "http/1.1"]`` to enable HTTP/2

Reviewer 2: The stdlib http.server does not support HTTP/2. Someone reading this will get the wrong impression.

   .. versionadded:: 3.7

.. class:: HTTPSServer(server_address, RequestHandlerClass,\
                       bind_and_activate=True, *, certfile, keyfile=None,\
                       password=None, alpn_protocols=None)

   Subclass of :class:`HTTPServer` with a wrapped socket using the :mod:`ssl` module.
   If the :mod:`ssl` module is not available, instantiating a :class:`!HTTPSServer`
   object fails with a :exc:`RuntimeError`.

   The *certfile* argument is the path to the SSL certificate chain file,
   and the *keyfile* is the path to file containing the private key.

   A *password* can be specified for files protected and wrapped with PKCS#8,
   but beware that this could possibly expose hardcoded passwords in clear.

>    .. seealso:: You can learn more about how SSL works because the *certfile*,
>    *keyfile* and *password* parameters are passed to the
>    :meth:`ssl.SSLContext.load_cert_chain`.

Reviewer: ```suggestion
   .. seealso::

      See :meth:`ssl.SSLContext.load_cert_chain` for additional
      information on the accepted values for *certfile*, *keyfile*
      and *password*.
```

Author: thanks

a *RequestHandlerClass* on instantiation, of which this module
provides three different variants:

> .. class:: HTTPSServer(server_address, RequestHandlerClass, \

Reviewer: This should be in the server paragraph, not in the handlers paragraph

>                        bind_and_activate=True, *, certfile, keyfile=None, \

Reviewer: You should mention that some parameters are forwarded to the superclass like `bind_and_activate` which is not in HTTPServer (why is it in HTTPSServer?)

Author: `HTTPServer` class also have this parameter because init method
`__init__(self, server_address, RequestHandlerClass, bind_and_activate=True)`
is inherited from `socketserver.TCPServer`. apparently they forgot to specify it in the documentation, I can fix it, if in your opinion this fix corresponds to the PR topic :-)

example:
```python
>>> inspect.signature(HTTPServer)
<Signature (server_address, RequestHandlerClass, bind_and_activate=True)>
```

Reviewer: Let's do that in a separate PR I'd say.

                       password=None, alpn_protocols=None)

>    This class is a :class:`HTTPServer` subclass with a wrapped socket using the

Reviewer: ALPN protocols should be explained or at least mentioned. We don't what kind lf type the argument should be, or its form.

   :mod:`ssl`, if the :mod:`ssl` module is not available the class will not
   initialize. The *certfile* argument is required and is the path to the SSL
   certificate chain file. The *keyfile* is the path to its private key. But
   private keys are often protected and wrapped with PKCS #8, so we provide
   *password* argument for that case.

>    .. versionadded:: 3.14

Reviewer: Use "next" instead of 3.14.

.. versionchanged:: 3.11
   Added the ``--protocol`` option.

The server can also support TLS encryption. The options ``--tls-cert`` and
``--tls-key`` allow specifying a TLS certificate chain and private key for

> secure HTTPS connections. And ``--tls-password`` option has been added to

Reviewer: ```
secure HTTPS connections. Use ``--tls-password`` option if private keys are passphrase-protected.
```

I'm on mobile so suggestions are hard to write. Btw, I think we say "passphrase-protected" not "password-protected".

``http.server`` to support password-protected private keys. For example, the
following command runs the server with TLS enabled::

>         python -m http.server --tls-cert cert.pem --tls-key key.pem

Reviewer: The example doesn't include the passphrase so it would be better to mention passphrase-protected keys after in a separate sentence (the "For example" would be otherwise confusing as it's not an example of the previous sentence)

.. versionchanged:: 3.11
   Added the ``--protocol`` option.

> The server can also support TLS encryption. The options ``--tls-cert`` and
> ``--tls-key`` allow specifying a TLS certificate chain and private key for
> secure HTTPS connections. Use ``--tls-password-file`` option if private keys are
> passphrase-protected. For example, the following command runs the server with
> TLS enabled::
> 
>         python -m http.server --tls-cert fullchain.pem
> 
> Or if a separate file with private key passphrase-protected::
> 
>         python -m http.server --tls-cert cert.pem --tls-key key.pem --tls-password-file password.txt
> 
> .. versionchanged:: next
>    Added the ``--tls-cert``, ``--tls-key`` and ``--tls-password-file`` options.

Reviewer: ```suggestion
The server can also support TLS encryption. The options ``--tls-cert`` and
``--tls-key`` allow specifying a TLS certificate chain and private key for
secure HTTPS connections. For example, the following command runs the server with
TLS enabled:

.. code-block:: bash

   python -m http.server --tls-cert fullchain.pem

Use ``--tls-password-file`` option if private keys are password-protected:

.. code-block::

   python -m http.server \
          --tls-cert cert.pem \
          --tls-key key.pem \
          --tls-password-file password.txt

.. versionchanged:: next
   Added the ``--tls-cert``, ``--tls-key`` and ``--tls-password-file`` options.
```

Ok, I'm sorry I said to change it to `passphrase` protected, but since we are naming the option `--tls-password-file`, I think it will be easier to remember that for readers if we use the same terminology (it's still correct though, but the openssl CLI usually asks for the *passphrase* not the *password*)

.. versionchanged:: 3.11
   Added the ``--protocol`` option.

The server can also support TLS encryption. The options ``--tls-cert`` and

> ``--tls-key`` allow specifying a TLS certificate chain and private key for
> secure HTTPS connections. For example, the following command runs the server with

Reviewer: ```suggestion
``--tls-key`` allow specifying a TLS certificate chain and a private key for
HTTPS connections. For example, the following command runs the server with
```

I think "secure HTTPS" is pleonasm but since the S in HTTPS is already for "secure". In addition, we don't necessarily want to say that the command-line setups the HTTPS server in a secure manner.

TLS enabled:

.. code-block:: bash

   python -m http.server --tls-cert fullchain.pem

> Use ``--tls-password-file`` option if private keys are password-protected:

Reviewer: ```suggestion
Use the ``--tls-password-file`` option if private keys are password-protected:
```

> .. code-block::

Reviewer: ```suggestion
.. code-block:: bash
```

.. versionchanged:: 3.11
   Added the ``--protocol`` option.

The server can also support TLS encryption. The options ``--tls-cert`` and
``--tls-key`` allow specifying a TLS certificate chain and private key for

> secure HTTPS connections. Use ``--tls-password`` option if private keys are

Reviewer 2: Passwords should never be put on a command line.  Don't add that flag unless it is named `--my-tls-password-is-not-secret-anymore-because-I-passed-it-on-a-command-line=` to tell people never to use it.

An appropriate way to get a password is from a file.  `--tls-password-file=` would be acceptable.

If you want to support use of `getpass`, make an explicit flag for just that.

   are not intended for use by untrusted clients and may be vulnerable
   to exploitation. Always use within a secure environment.

.. option:: --tls-cert

>    The server can also support TLS encryption. The option ``--tls-cert`` allow
>    specifying a TLS certificate chain for HTTPS connections. For example,
>    the following command runs the server with TLS enabled::
> 
>       python -m http.server --tls-cert fullchain.pem
> 
>    .. versionadded:: next

Reviewer: ```suggestion
   Specifies a TLS certificate chain for HTTPS connections:

      python -m http.server --tls-cert fullchain.pem

   .. versionadded:: next
```

.. option:: --tls-key

>    Specifies private key for HTTPS connections.

Reviewer: ```suggestion
   Specifies a private key file for HTTPS connections.

   This option requires ``--tls-cert`` to be specified.
```

   .. versionadded:: next

> .. option:: --tls-password-file
> 
>    Use the ``--tls-password-file`` option if private keys are
>    password-protected::
> 
>       python -m http.server \
>              --tls-cert cert.pem \
>              --tls-key key.pem \
>              --tls-password-file password.txt
> 
>    .. versionadded:: next

Reviewer: ```suggestion
.. option:: --tls-password-file

   Specifies the password file for password-protected private keys:

      python -m http.server \
             --tls-cert cert.pem \
             --tls-key key.pem \
             --tls-password-file password.txt

   This option requires `--tls-cert`` to be specified.

   .. versionadded:: next
```
